utils/preferences/src/macros.rs

```diff
@@ -0,0 +1,148 @@
+#[macro_export]
+macro_rules! preferences {
+    ($name:ident,
+     $default_name:ident,
+     $resolved_name:ident,
+     {$($key:ident => $pref:ty, $resolved:ty, $ue:expr),*},
+     $options_name:ident,
+     {$($k:ident => $v:ty),*}
+     ) => (
+        #[derive(Default)]
+        #[non_exhaustive]
+        pub struct $name {
+            pub lid: Option<icu_locid::LanguageIdentifier>,
+            $(
+                pub $key: $pref,
+            )*
+        }
+
+        #[non_exhaustive]
+        pub struct $default_name {
+            $(
+                pub $key: $resolved,
+            )*
+        }
+
+        pub struct $resolved_name {
+            pub lid: icu_locid::LanguageIdentifier,
+
+            $(
+                pub $key: $resolved,
+            )*
+        }
+
+        impl From<(LanguageIdentifier, &$default_name)> for $resolved_name {
+            fn from(input: (LanguageIdentifier, &$default_name)) -> Self {
+                Self {
+                    lid: input.0,
+                    $(
+                        $key: input.1.$key,
+                    )*
+                }
+            }
+        }
+
+        impl Preferences for $name {
+            fn language(&self) -> &icu_locid::subtags::Language {
+                self.lid
+                    .as_ref()
+                    .map(|lid| &lid.language)
+                    .unwrap_or(&icu_locid::subtags::Language::UND)
+            }
+
+            fn script(&self) -> Option<&icu_locid::subtags::Script> {
+                self.lid.as_ref().and_then(|lid| lid.script.as_ref())
+            }
+
+            fn region(&self) -> Option<&icu_locid::subtags::Region> {
+                self.lid.as_ref().and_then(|lid| lid.region.as_ref())
+            }
+        }
+
+        impl From<Locale> for $name {
+            fn from(loc: Locale) -> Self {
+                let lid = Some(loc.id);
+
+                $(
+                    let mut $key = None;
+                )*
+
+                for (k, v) in loc.extensions.unicode.keywords.iter() {
+                    $(
+                      if let Some(ue) = &$ue {
+                          if k == ue {
+                              if let Ok(r) = TryInto::try_into(v) {
+                                  $key = Some(r);
+                              }
+                          }
+                      }
+                    )*
+                }
+
+                Self {
+                    lid,
+                    $(
+                        $key,
+                    )*
+                }
+            }
+        }
+
+        impl $name {
+            #[allow(clippy::result_unit_err)]
+            pub fn merge_locale(&mut self, locale: &Locale) -> Result<(), ()> {
+                if let Some(lid) = &mut self.lid {
```

Comment:
Nit: Move the first part of this function into `LanguageIdentifier::merge(&mut self, other: LanguageIdentifier)` and get it out of the macro. The less code inside the macro the better, for numerous reasons I can list out if you disagree.